### ISSUE TITLE
--line-numbers maybe?

### ISSUE THREAD

**Issue by goors** [ldx/python-iptables#161]
I can not seam to find line numbers in my rules. It would be great if you can add that. 

The when I want to delete I do not have to match everything that i added before but just to do simple delete by line number.

I mean all of this code:

    >>> import iptc
    >>> table = iptc.Table(iptc.Table.FILTER)
    >>> table.autocommit = False
    >>> chain = iptc.Chain(table, "FORWARD")
    >>> for rule in chain.rules:
    >>>     if rule.out_interface and "eth0" in rule.out_interface:
    >>>         chain.delete_rule(rule)
    >>> table.commit()
    >>> table.autocommit = True

to delete just one rule

**Comment by ldx** [comment]
Thanks for the suggestion, I'll look into this.

**Comment by tomislacker** [comment]
+1

Just ran into this myself in a similar situation.  Was looking to output a table of rule number, target, and counters for diagnostics.  I started playing around in the code but haven't gotten anywhere.  I'll send up a push if I do.

**Comment by ldx** [comment]
@goors @tomislacker can you guys show me an example how this would ideally work?

**Comment by goors** [comment]
I am not sure that I understand you correctly. Lets say you have something like this:

<pre>$bash: iptables -L DOCKER -n --line-numbers</pre>

That will list DOCKER chain like this:

<pre>
Chain DOCKER (1 references)
num  target     prot opt source               destination
1    ACCEPT     tcp  --  0.0.0.0/0            172.17.0.1           tcp dpt:9002
2    ACCEPT     tcp  --  0.0.0.0/0            172.17.0.2           tcp dpt:9001
3    ACCEPT     tcp  --  0.0.0.0/0            172.17.0.2           tcp dpt:9000
</pre>

And then to delete, you will have something like:

<pre>
iptables -D DOCKER 3
</pre>

And in code:

<pre>
chain = iptc.Chain(table, "DOCKER")
chain.rules[2].delete() # something like this or maybe some getter to get rule by id and delete it?
</pre>

**Comment by ldx** [comment]
This should work:

    table = iptc.Table(iptc.Table.FILTER)
    chain = iptc.Chain(table, "FORWARD")
    chain.delete_rule(chain.rules[1])

The list `chain.rules` is zero-indexed, though, while `--line-numbers` in `iptables` is 1-indexed.

Does this help?

**Comment by tomislacker** [comment]
@ldx I think we all realized we could access the rules like that.  I'm not sure if @goors would agree, but my interest in this issue came along due to the lack of clarity in making the assertion that your above code example makes. *(`rule[0]` -> "Line 1", `rule[1]` -> "Line 2", etc..)*  It's not the fault of the python-iptables project, but rather the iptables project.

Given that [_get_rules()](../blob/4f4dea955dcfd754ace1718b290cdf9493647d75/iptc/ip4tc.py#L1459) calls `iptc_next_rule` *(implemented as `TC_NEXT_RULE` in the latest code)*, it seems that we should be able to safely make the assumption.  However, it'd be nice if the iptables folks could shore up this concern by adding a little bit of explicit documentation to the code that indicates rules will always be produced in order:

```c
/* Returns NULL when rules run out. */
const STRUCT_ENTRY *
TC_NEXT_RULE(const STRUCT_ENTRY *prev, struct xtc_handle *handle)
{
    struct rule_head *r;

    iptc_fn = TC_NEXT_RULE;
    DEBUGP("rule_iterator_cur=%p...", handle->rule_iterator_cur);

    if (handle->rule_iterator_cur == NULL) {
        DEBUGP_C("returning NULL\n");
        return NULL;
    }

    r = list_entry(handle->rule_iterator_cur->list.next,
            struct rule_head, list);

    iptc_fn = TC_NEXT_RULE;

    DEBUGP_C("next=%p, head=%p...", &r->list,
        &handle->rule_iterator_cur->chain->rules);

    if (&r->list == &handle->rule_iterator_cur->chain->rules) {
        handle->rule_iterator_cur = NULL;
        DEBUGP_C("finished, returning NULL\n");
        return NULL;
    }

    handle->rule_iterator_cur = r;

    /* NOTE: prev is without any influence ! */
    DEBUGP_C("returning rule %p\n", r);
    return r->entry;
}
```

**tl;dr** I'd say we can close the issue but for my own sanity, I may write a smoke test that assures me that each combination of the kernel+libiptc that my code runs on still enables this assumption.

**Comment by ldx** [comment]
@tomislacker as far as I know, that assumption is correct, rules are always given back in order and are 1-indexed from the viewpoint of iptables.

Unfortunately, python-iptables could only add rule number information to the rule objects via quite some extra runtime overhead (always traversing all the rules in the chain to make sure numbers are not off, e.g. when rules have been added/removed in another process).

Instead, I'd opt for adding more documentation on how rules are numbered. Feel free to send a PR, not sure when I'll have the time to do it myself.

**Comment by tomislacker** [comment]
> Unfortunately, python-iptables could only add rule number information to the rule objects via quite some extra runtime overhead (always traversing all the rules in the chain to make sure numbers are not off, e.g. when rules have been added/removed in another process).

Yea I'm tracking, that's about the conclusion I came to as well and I don't personally think it's worth it.  I'm of the opinion that for the use-case of enumerating all rules in a chain, using the list index is sufficient.  And for all other use cases, the developer should be mindful *not* to use rule numbers as they cannot be positively identified without additional overhead *(which they could program themselves)*.

As far as documentation, we could at least denote that we're at the mercy of libiptc.  What do you think about tossing in a test that could be run manually to assert the rule ordering?

**Comment by ldx** [comment]
> As far as documentation, we could at least denote that we're at the mercy of libiptc. What do you think about tossing in a test that could be run manually to assert the rule ordering?

Sure, sounds like a good idea. Can you send a PR?
